virttest/qemu_devices/qbuses.py

```diff
@@ -650,6 +650,118 @@ def _increment_addr(self, addr, last_addr=None):
+        """
+        Justify if the device should be plug to pcie bus directly,
+        the device is not directly plugged only if it is a virtio
+        pci device and the device param direct_plug is "no".
+
+        :param device: the QBaseDevice object
+        :return: bool value for directly plug or not.
+        """
+        pcie_devices = ["virtio-net-pci", "virtio-blk-pci",
+                        "virtio-scsi-pci", "virtio-balloon-pci",
+                        "virtio-serial-pci", "virtio-rng-pci",
+                        "e1000e", "virtio-gpu-device"]
+        if device.get_param("driver") in pcie_devices and (
+                    device.get_param("direct_plug_pci", "no") == "no"):
+            return False
+
+        return True
+
+    def add_root_port(self, addr, root_port=None, multifunction="off"):
+        """
+        Add pcie root port to the bus and __root_ports list, assign free slot,
+        chassis, and addr for it according to current list.
+
+        :param addr: addr to add the root port
+        :param root_port: pcie-root-port QDevice object
+        :param multifunction: if multifunction on or off
+        :return: pcie-root-port QDevice object
+        """
+        _addr = int(addr, 16)
+        if not root_port:
+            root_port = self._add_root_port(addr, multifunction)
+        if _addr not in self.__root_ports:
+            bus = root_port.child_bus[0]
+            params = {'slot': _addr, 'chassis': _addr, 'addr': "0x%x" % _addr}
+            for key, value in params.iteritems():
+                root_port.set_param(key, value)
+            self.__root_ports["0x%x" % _addr] = bus
+        return root_port
+
+    def _add_root_port(self, addr, multifunction="off"):
+        """
+        Generate pcie-root-port QDevice object with certain addr and id
+
+        :param addr: addr to add the root port
+        :param multifunction: if multifunction on or off
+        :return: pcie-root-port QDevice object
+        """
+        _addr = int(addr, 16)
+        if _addr not in self.__root_ports:
+            bus_id = "%s-root-port-%s" % (self.busid, _addr)
+            length = 1 if multifunction == "off" else 8
+            bus = QPCIBus(bus_id, 'PCIE', bus_id, length=length)
+            params = {'id': bus_id}
+            return qdevices.QDevice(self.__root_type,
+                                    params,
+                                    aobject=bus_id,
+                                    parent_bus=[{'aobject': self.aobject}],
+                                    child_bus=bus)
+
+    def _insert(self, device, addr):
+        """
+        Inserts a pcie-root-port before the device if it is not a
+        direct plug device.
+
+        :param device: qdevice.QDevice object
+        :param addr: addr to insert the device
+        """
+        _addr = addr.split('-')[0]
+        if self.is_direct_plug(device):
+            if device.get_param("driver") == self.__root_type:
```

Comment:
The question is whether this should be only for the `self.__root_type`, or another whitelist of `root-port` devices. I think the later would be better as it'd allow one to mix-and-match different root-port types and use them the same way.

Reply:
@luckyh @ldoktor Thanks. In pcie test, ioh3420 should never appear together with pcie-root-port, it is just a transition name,and will be replaced by pcie-root-port totally later. We only keep it for old version adapt, so control it by root_port_type() function in qcontainer.DevContainer, to make sure only one of them used.

Reply:
I should translate ioh3420 and pcie-root-port that specify by user to correct supported one in qcontainer.DevContainer.pcic_by_params()

Reply:
Well the `ioh3420` is implemented according the intel driver, while the `pcie-root-port` is a fresh clean implementation. In the future it might happen (unlikely) that someone creates another implementation and we might want to test combination of some of iohs and some pcie-rp.

Anyway this is not a strong opinion, I think it'd be nicer to whitelist all of the root-ports here, but I can live with this implementation as well. On the other hand I don't think you should translate `ioh3420` to `pcie-root-port` as people might want to specify they really want to test `ioh`s.